### PR TITLE
feat(server): support esm

### PR DESCRIPTION
**Pull request by yimingjfe**
###  How can we make projects run in esm format?

There are two options for Modern.js to support application project code running in esm format without generating a break change:
1. Transformed Modern.js packages to dual package, that is, the whole package using node.js conditional exports, providing two different exports, the problem of this solution is that the transformation cost is high (you need to transform all packages at a time, at present, many of our packages are supported by different exports, but in fact, the code is not supported, for example, still using ` require`, `__dirname` APIs that are only available in commonjs module scope), and the potential for multi-instance problems (since the dependencies in the application project may be imported from two different exports)

2. Using the features that the esm project  can import the commonjs module. The Modern.js packages actually runs in the as commonjs, but because Modern.js requires the source code of the `require` user, commonjs cannot import esm module sync, so we modify the `require` function of Modern.js for user code to support import commonjs module and esm module both.

Here we use the second way so that the user's project can run as esm format. At the same time, we can gradually transform the existing package to reduce the risk of transformation.

### How can we make ts projects run in esm format?

In the commonjs period, the `ts-node` tools worked by modifying `require.extensions`, but `require.extensions` doesn't support esm modules. node.js supports custom loaders via `node --loader ts-node/esm`, but the problem is that we need to support ts paths, and now both `ts-node` and `tsconfig-paths` don't support paths for ts in the esm projects. so here we utilize node.js's [hooks](https://nodejs.org/api/module.html#hooks) and tsconfig-paths to support the framework's built-in aliases. (However, due to the limitations of the tsconfig-paths implementation, it is not possible to support user-configured paths in the project, and there are other tools that can support user-configured paths in the`tsconfig.json` but cannot support the framework's custom built-in aliases.

### Other questions：

1. Why are there `.mjs` files in `app-tools` instead of using `.mts`?

Our solution essentially runs Modern.js packages in the commonjs format, `@modern-js/app-tools` is built using module-tools, and after `module-tools` builds, files with `.mts` extensions are compiled to `.js` extensions, which makes these esm files will be compiled into `.js` files, making these esm files run in `commonjs` format.

2. Why are `require.extensions` and `hooks` registered at the same time?

Because in esm projects, you have to compile both ts code in esm format and ts code in commonjs format, and for esbuild register, esbuild/loader actually relies on esbuild register (which converts the code and marks it as  [commonjs format](https:/ /github.com/egoist/esbuild-register/blob/dev/src/loader.ts)

3. What was changed in the entry for deployment?

All projects build `cjs` entries, and at the same time, esm projects build esm entries, which import `cjs` entries through `esm` entries.

### Main changes

1. Modify the package artifacts fomat to avoid compiling `await import` into `require`.
5. Add a loader in app-tools by the node.js Module's custom hooks, and modify the original register logic and adapt it to projects that use esbuild.
6. Modify logic of server-side aliases
7. Modify the format of SSR bundle
8. Modify the entry of deployment
9. Converge the require to compatiableRequire in the framework, and change the function that uses it to async
10. Add some types to some package.json
11. Remove the `file-extension-in-import`rule

## Checklist

- [ ] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify-handler.cjs
Comment (zllkjc) [inline packages/solutions/app-tools/src/plugins/deploy/platforms/netlify-handler.cjs:1]: 这个文件一定要用 .cjs 命名吗

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts [hunk 1]
```diff
@@ -114,22 +124,34 @@ export const createNetlifyPreset: CreatePreset = (
         })
         .join(',')}]`;
 
-      let entryCode = (
-        await fse.readFile(path.join(__dirname, './netlifyEntry.js'))
+      let handlerCode = (
+        await fse.readFile(path.join(__dirname, './netlify-handler.js'))
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts:128]: 这里为啥又是 js 呢

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts [hunk 2]
```diff
@@ -114,22 +124,34 @@ export const createNetlifyPreset: CreatePreset = (
...
       ).toString();
 
       const serverAppContext = serverAppContenxtTemplate(appContext);
 
-      entryCode = entryCode
-        .replace('p_genPluginImportsCode', pluginImportCode)
+      handlerCode = handlerCode
+        .replace('handlerCode', pluginImportCode)
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts:134]: 我看 netlify-handler 里还是 p_genPluginImportsCode

File: packages/solutions/app-tools/src/plugins/deploy/platforms/vercel-handler.cjs
Comment (zllkjc) [inline packages/solutions/app-tools/src/plugins/deploy/platforms/vercel-handler.cjs:1]: 这个为啥和 netlify-entry.cjs 的导出方式不同，一个是 CJS 一个 ESM

File: packages/solutions/app-tools/src/plugins/deploy/platforms/vercel.ts [hunk 3]
```diff
@@ -137,7 +146,19 @@ export const createVercelPreset: CreatePreset = (
...
+      await fse.writeFile(handlerFilePath, handlerCode);
+      if (isEsmProject) {
+        // We will not modify the entry file for the time, because we have not yet converted all the packages available for esm.
+        await fse.copy(
+          path.join(__dirname, './vercel-entry.mjs'),
+          entryFilePath,
+        );
+      } else {
+        await fse.copy(
+          path.join(__dirname, './vercel-entry.js'),
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/plugins/deploy/platforms/vercel.ts:158]: 看起来没有这个文件？

File: packages/solutions/app-tools/src/utils/config.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ export const buildServerConfig = async ({
       `${filepath.replace(
         new RegExp(CONFIG_FILE_EXTENSIONS.join('|')),
         '',
-      )}.js`,
+      )}.cjs`,
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/utils/config.ts:38]: 这个改动的目的是？

File: packages/solutions/app-tools/src/utils/register.ts [hunk 5]
```diff
@@ -40,33 +69,43 @@ export const registerCompiler = async (
...
+    await registerEsbuild({
+      isTsProject,
+      tsConfig,
+      distDir,
     });
   }
 
-  const tsConfigPaths = await import('@modern-js/utils/tsconfig-paths');
+  const tsConfigPaths = (await import('@modern-js/utils/tsconfig-paths'))
+    .default;
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/utils/register.ts:108]: 这个改动是什么原因，之前的 bug？

File: packages/toolkit/runtime-utils/modern.config.js [hunk 6]
```diff
@@ -1,7 +1,21 @@
 const { universalBuildConfig } = require('@scripts/build');
 
+universalBuildConfig[0].esbuildOptions = options => {
+  options.supported = {
```
Comment (zllkjc) [inline packages/toolkit/runtime-utils/modern.config.js:4]: 同样的问题，多份一样的配置

File: tests/integration/bff-koa/modern.config.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
-import AppToolsPlugin, { defineConfig } from '@modern-js/app-tools';
-import bff from '@modern-js/plugin-bff';
-import koa from '@modern-js/plugin-koa';
+import { appTools, defineConfig } from '@modern-js/app-tools';
+import { bffPlugin } from '@modern-js/plugin-bff';
+import { koaPlugin } from '@modern-js/plugin-koa';
```
Comment (zllkjc) [inline tests/integration/bff-koa/modern.config.ts:3]: 开发者也需要做这样的改动吗？

File: tests/integration/pure-esm-project/modern.config.js
Comment (zllkjc) [inline tests/integration/pure-esm-project/modern.config.js:1]: 是否需要添加一个 ts 的用例呢？

File: packages/toolkit/utils/src/cli/require.ts [hunk 8]
```diff
@@ -2,15 +2,33 @@ import { findExists } from './fs';
...
  */
-export const compatRequire = (filePath: string, interop = true) => {
-  const mod = require(filePath);
-  const rtnESMDefault = interop && mod?.__esModule;
+export async function compatibleRequire(
+  path: string,
+  interop = true,
+): Promise<any> {
+  if (path.endsWith('.json')) {
+    return require(path);
```
Comment (zllkjc) [inline packages/toolkit/utils/src/cli/require.ts:13]: 这里为什么不需要 import

File: packages/toolkit/utils/src/cli/require.ts [hunk 9]
```diff
@@ -2,15 +2,33 @@ import { findExists } from './fs';
...
+  }
 
-  return rtnESMDefault ? mod.default : mod;
-};
+  let requiredModule;
+  try {
+    requiredModule = require(path);
+  } catch (err: any) {
+    if (err.code === 'ERR_REQUIRE_ESM' || err instanceof SyntaxError) {
```
Comment (zllkjc) [inline packages/toolkit/utils/src/cli/require.ts:20]: 如果是 type: module 的情况下，感觉应该直接使用 await import？

File: packages/toolkit/utils/src/cli/require.ts (hunk 8 above)
Comment (yimingjfe) [reply packages/toolkit/utils/src/cli/require.ts:13]: 因为不是 js 呀

File: tests/integration/pure-esm-project/modern.config.js
Comment (yimingjfe) [reply tests/integration/pure-esm-project/modern.config.js:1]: 我把 bff-koa 那里改为 ts 的用例了，有 ssr，也有 bff 的

File: packages/solutions/app-tools/src/utils/register.ts (hunk 5 above)
Comment (yimingjfe) [reply packages/solutions/app-tools/src/utils/register.ts:108]: 之前这里被编译为 `require` 了，现在走的 `import`，所以加了 default

File: packages/solutions/app-tools/src/utils/config.ts (hunk 4 above)
Comment (yimingjfe) [reply packages/solutions/app-tools/src/utils/config.ts:38]: 产物是 cjs 格式的，改为 `.cjs` 后缀，避免在 esm 项目中被认为是 `esm` 格式的代码

File: packages/solutions/app-tools/src/plugins/deploy/platforms/vercel.ts (hunk 3 above)
Comment (yimingjfe) [reply packages/solutions/app-tools/src/plugins/deploy/platforms/vercel.ts:158]: done~

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts (hunk 2 above)
Comment (yimingjfe) [reply packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts:134]: 这里应该是改错了，done~

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts (hunk 1 above)
Comment (yimingjfe) [reply packages/solutions/app-tools/src/plugins/deploy/platforms/netlify.ts:128]: cjs 的为了保证在 esm 项目中运行正常，`.js` 后缀的文件不会出现在 esm 项目中

File: packages/solutions/app-tools/src/plugins/deploy/platforms/netlify-handler.cjs
Comment (yimingjfe) [reply packages/solutions/app-tools/src/plugins/deploy/platforms/netlify-handler.cjs:1]: 是的，`cjs` 指明这个文件的模块是 commonjs 模块，不管是 esm 项目还是 commonjs 项目都能得到正确的处理

File: packages/server/bff-core/modern.config.js [hunk 10]
```diff
@@ -1,5 +1,21 @@
 const { nodeBuildConfig } = require('@scripts/build');
 
+nodeBuildConfig[0].esbuildOptions = options => {
+  options.supported = {
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/server/bff-core/modern.config.js:4]: 一个是编译为 cjs 产物的配置，一个编译 esm-node 产物的配置，都需要支持 dynamic import

File: packages/runtime/plugin-runtime/src/cli/ssr/index.ts [hunk 11]
```diff
@@ -136,6 +137,20 @@ export const ssrPlugin = (): CliPlugin<AppTools> => ({
           },
           tools: {
             babel: babelHandler,
+            bundlerChain: (chain, { isServer }) => {
+              if (isServer && appContext.moduleType === 'module') {
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/runtime/plugin-runtime/src/cli/ssr/index.ts:141]: 总体上，之前的 ssr bundle 的产物是 commonjs， server 会 require 这个 commonjs 模块，当项目的 package.json 中 type 配置为 `module`，`bundle.js` 会被当做 esm 模块，所以这里为了构建出 `esm` 模块，这块其实还挺清晰的~

File: packages/solutions/app-tools/modern.config.js [hunk 12]
```diff
@@ -1,5 +1,72 @@
+const path = require('path');
 const { universalBuildConfig } = require('@scripts/build');
 
+universalBuildConfig[0] = {
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/solutions/app-tools/modern.config.js:4]: 所有 `modern.config.js` 这里的改动都是为了避免 dynamic import 被编译 commonjs 的 `require`，这里为了覆盖默认配置

File: packages/cli/plugin-bff/src/cli.ts [hunk 13]
```diff
@@ -126,21 +118,14 @@ export const bffPlugin = (): CliPlugin<AppTools> => ({
         return { plugins };
       },
 
-      async beforeBuild() {
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/cli/plugin-bff/src/cli.ts:129]: 是的，这里之前是处理 bff 别名的逻辑，依赖了 `Module.__resolvePathname`， 目前把内置别名的处理逻辑收敛到 `packages/solutions/app-tools/src/utils/register.ts` 中了

File: packages/cli/core/src/types/context.ts [hunk 14]
```diff
@@ -45,6 +45,8 @@ export interface IAppContext {
   port?: number;
   /** Name of the current project's package.json */
   packageName: string;
+  /** Type field of package.json */
+  moduleType: 'module' | 'commonjs';
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/cli/core/src/types/context.ts:49]: 默认改为 commonjs 了，与 node.js 包行为保持一致

File: packages/toolkit/utils/src/cli/require.ts (hunk 9 above)
Comment (yimingjfe) [reply packages/toolkit/utils/src/cli/require.ts:20]: 不太行，这里有些场景下还是会 require commonjs 的代码，比如我们的 server 配置

File: packages/solutions/app-tools/modern.config.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/solutions/app-tools/modern.config.js:4]: 一个是编译为 cjs 产物的配置，一个编译 esm-node 产物的配置，都需要支持 dynamic import

File: packages/solutions/app-tools/src/esm/esbuild-loader.mjs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    alias,
+    appDir,
+    tsconfigPath,
+  });
+}
+
+export function resolve(specifier, context, defaultResolve) {
+  const match = matchPath(specifier);
+  return match
+    ? defaultResolve(pathToFileURL(match).href, context)
```
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/solutions/app-tools/src/esm/esbuild-loader.mjs:16]: 对的，alias 的逻辑，为了能走 esm 的 resolve，这里要把路径通过 `pathToFileURL` 修改为文件协议的 url

[the remark this answers is not in the document]
Comment (zllkjc) [reply packages/solutions/app-tools/src/esm/esbuild-loader.mjs:16]: 这个有啥文档介绍不

File: packages/toolkit/utils/src/cli/require.ts (hunk 9 above)
Comment (zllkjc) [reply packages/toolkit/utils/src/cli/require.ts:20]: OK，这里标记下，现在相当于最小改动。社区里有类似的实现吗。

File: packages/solutions/app-tools/src/esm/esbuild-loader.mjs (hunk 15 above)
[the remark this answers is not in the document]
Comment (yimingjfe) [reply packages/solutions/app-tools/src/esm/esbuild-loader.mjs:16]: cc  https://nodejs.org/api/esm.html#import-specifiers
image

**Review by zllkjc — APPROVED** [review]

**Review by zllkjc — APPROVED** [review]

**Pull request merged**